### ISSUE TITLE
getJSON fetch failure should not abort build

### ISSUE THREAD

**Issue by anthonyfok** [gohugoio/hugo#5643]
Hugo >= 0.50 (including master HEAD as of today) exits before finishing building a website when getJSON fails to get a resource.

This problem was earlier filed as Issue #5076 "getJSON etc. should log ERROR but return nil in failure situations" and was fixed in 0.49, but perhaps the "greatly improved error messages shown in browser" feature in 0.50 might have undone the fix in PR #5199.

**Comment by anthonyfok** [comment]
Result of `git bisect v0.50 v0.49`:

d1661b823af25c50d3bbe5366ea40a3cdd52e237 is the first bad commit

```
commit d1661b823af25c50d3bbe5366ea40a3cdd52e237
Author: Bjørn Erik Pedersen <bjorn.erik.pedersen@gmail.com>
Date:   Sun Oct 21 12:20:21 2018 +0200

    hugolib: Continue the file context/line number errors work
    
    See #5324

:040000 040000 38ca8acc76fdeb7e60690e791fff57a70554da77 2688e5c2e7a48f5d98a7064acba7177c818c1b62 M	commands
:040000 040000 d7864d4c82c9cd7f06c7c9e9da50eb0415550a40 3682f042b8bfa5bcbe2d7d67b9fcc335af6ea7c9 M	common
:040000 040000 f2695804be8598aeacb913cd0b39aed748b6cf0c d2a88525f51ed4afaefb268190668d5dd20c704f M	deps
:040000 040000 81458ba1b9871204f15e511a99990023e4b169c6 6debd34332af22bed2e388bd16d5bea97abff045 M	hugolib
:040000 040000 3b2b2ef86859a71a7634fe69f7e3ad46cc06e358 c101dd1acc12816b7d698a6c28a5379236586dd5 M	parser
:040000 040000 efc60361fcbdbb3c9d1020826a54de1b675737e9 07f18b7c9e70b051ee807258d43def3cb14cee66 M	tpl
```

**Comment by bep** [comment]
Not sure what finding how finding who/what broke it will help fixing it. Why not just fix it?

**Comment by anthonyfok** [comment]
I am actually trying to fix it; I am just documenting the progress along the way by pasting the result of the last "git bisect" command, though probably the output of that command was too verbose. I'll edit my last message to remove the details and just keep the commit hash to keep track of it.

**Comment by anthonyfok** [comment]
Trying to find clues from the messages shown by `hugo -v`...

Before:

```
ERROR 2019/01/27 16:16:23 Failed to get JSON resource https://api.example.com/product?name=foobar: Failed to retrieve remote file: Not Found
Total in 1234 ms
Error: Error building site: logged 1 error(s)
```

After:

```
Error: Error building site: failed to render pages: [en] page "Test Website": render of "home" failed: execute of template failed: template: index.html:24:13: executing "main" at <getJSON "https://api...>: error calling getJSON: failed to get getJSON resource "https://api.example.com/product?name=foobar": Failed to retrieve remote file: Not Found
```

**Comment by anthonyfok** [comment]
My first attempt at a fix is at PR #5648, though I am starting to think I tried to fix the problem at the wrong place, and at a cost too: The nice context information provided by _errors.Wrapf() (from pkg/errors) is lost.  So, simply replacing with _errors.Wrapf() with ns.deps.Log.ERROR.Printf() reverting to the pre-0.50 behaviour of no error context isn't ideal.

I am thinking that there should be a way to distinguish between fatal and non-fatal errors, maybe with the non-fatal ones marked with a "[non-fatal]" suffix in the error string so it can be checked later.  (I have just barely started looking at the error handling code, so what I just said may be utter nonsense, my apologies.)  Unfortunately, I won't have time to look at this issue again in this week or the next, hence this note in case anyone else (or maybe the future "me") wants to pick it up.

**Comment by Lnaden** [comment]
I'd like to revive this issue as failures in `getJSON` still cause builds to fail and halt in the versions 0.68.3, 0.69.1 and 0.69.2 which I have tested it on. 

The use case I have is that we're checking a database through its REST API for the existence of an entry to pull some metadata, and falling back to local data if not present. However, the only way we can check for the existence is to make a query, and if it's not there, the `getJSON` errors, and the build halts.

Specific error is this: `Failed to get JSON resource "https://data.rcsb.org/rest/v1/core/entry/": Failed to retrieve remote file: Not Found`. 

Supposedly #5648 was meant to fix this, but from what I'm reading never did it properly? 

I don't know Go so I'm not sure exactly how to fix it, but I think I can trace its stack:
* `getJson` calls `getResource` and then if an error is thrown its supposed to log said error but _not throw_ the error: https://github.com/gohugoio/hugo/blob/c2d9fd1ebef89cc19256e6799f29ab3d27eaf879/tpl/data/data.go#L118-L122
* `getResource` attempts to call `getRemote` because the a URL was passed in, so the `""` case is skipped and it goes to the `default`: https://github.com/gohugoio/hugo/blob/c2d9fd1ebef89cc19256e6799f29ab3d27eaf879/tpl/data/resources.go#L109-L122
* `getRemote` detects an `HTTPError` and _actually throws_ the error, which passes back up the stack: https://github.com/gohugoio/hugo/blob/c2d9fd1ebef89cc19256e6799f29ab3d27eaf879/tpl/data/resources.go#L55-L57
* `getResource` and `getJSON` behave correctly and log the error, but the thing actually halting the build came from the `getRemote` call and the `HTTPError` handling.

I have a couple ideas on what behavior I would like to see to make this more user and use-case friendly.

1. Is it possible to make the `getRemote` on `HTTPError` not actually throw the error and instead log it, pass it back up and let the calling function handle it? Especially since `getRemote` isn't a publicly available function in Hugo's shortcodes?
1. Could there be some switch to tell the `getRemote` to pass it back up (like if there is another function invoking which could handle it?)
1. Could the `getRemote` or `getJson` throw a warning instead? This does leave it up to the site developer to properly handle the case when the JSON they get back isn't what they expect, but I think that's a reasonable thing to ask for.
2. Use a `with...else...end` construct to handle the return from `getJSON` when it errors. This actually already works as expected! I can test it with the shortcode construct of  
    ```
    {{ with getJson "...." }}
        <!-- Do stuff on success -->
    {{ else }}
        {{ warnf "It failed" }}
    {{ end }}
    ```
    This will correctly throw not only the `getJSON` error, but also throw the `warnf` warning in the console. However, the fact that an error was thrown causes the built to fail in the end, but the `getJSON` error clearly is not fatal to the overall build code.
3. Have a function to probe an external site? Basically a function to see if the url exists. I think `getRemote` does this already to some extent, but make this return a boolean instead of the result or error?

**Comment by pgorod** [comment]
If I may just add my use case, to explain why this is important to me. This is from a forums thread a while ago:

> An example just to try to get you to sympathize: please see [this page](https://docs.suitecrm.com/admin/releases/7.11.x/) on our site. Scroll down until you see the Github avatars of the contributors. That’s coming form the Github API.
>
> So, one day it’s working. A few months later, after some totally unrelated change in content, the Netlify build is failing because of the getJSON call. The cause? One of the contributors deleted his account on Github, and the call to get his Avatar fails.
>
> Obviously, I should just be able to put in a dummy placeholder Avatar and get on with my life. An API is for consuming dynamic information. A query that returns 0 results is not necessarily an application error - it can be a normal every day occurrence, and we need to be able to handle it.

@Lnaden thanks for reviving this and for your thoughts on the subject.

**Comment by anthonyfok** [comment]
Hi @Lnaden,

Thank you for your report.  However, since this issue was already fixed on v0.54,
what you see may be similar manifestation of a different cause in later version.

So, instead of "reviving" an old fixed issue, I think it would be better to open up a new issue that mentions this old issue.

Thanks!
